File: waite/stage2/iwch.f
      SUBROUTINE IWRCH(JCHAR,JBUFF,JINDX,JFLAG,JLENG) 
C 
C  THIS ROUTINE ENTERS A CHARACTER INTO THE LINE BUFFER.  IF THE
C  CHARACTER IS NEGATIVE, IT IS TAKEN TO BE A TERMINATOR.  IF THE LINE
C  BUFFER IS FULL, IT IS ALSO TERMINATED AND THE CHARACTER IS IGNORED.
C  IF THE LINE BUFFER IS TERMINATED, A FLAG IS RETURNED TO THE USER 
C  INDICATING THIS FACT.  THE LINE BUFFER LIMIT IS SET TO INDEX THE 
C  POSITION BEYOND THE LAST SIGNIFICANT CHARACTER.  IF THE LINE BUFFER
C  IS NOT TERMINATED, THEN THE CHARACTER IS ENTERED AT THE INDEXED
C  POSITION AND THE INDEX IS INCREMENTED.  THE LIMIT INDEX IS NOT 
C  ALTERED IN THIS CASE.
C 
C         JCHAR  IS THE INTEGER REPRESENTING THE CHARACTER
C 
C         JBUFF  IS THE ADDRESS OF THE LINE BUFFER
C 
C         JINDX  IS THE INDEX TO THE LINE BUFFER
C 
C         JFLAG IS SET TO ZERO IF THE LINE BUFFER IS NOT FULL AND THE 
C              CHARACTER INSERTED WAS NOT CR/LF.  IT IS SET TO 1
C              OTHERWISE
C 
C        JLENG IS THE LAST+1 INDEX OF INFORMATION IN THE BUFFER 
C 
      DIMENSION JBUFF(81) 
C 
C         IF THE CHARACTER IS NEGATIVE, THE BUFFER MUST BE TERMINATED.
C         IT IS IMMATERIAL WHETHER THE BUFFER IS FULL OR NOT. 
C 
      IF (JCHAR .LT. 0)  GO TO 1
C 
C        IF THE USER IS ATTEMPTING TO PUT A SIGNIFICANT CHARACTER INTO
C         THE LINE BUFFER, WE MUST ENSURE THAT THERE IS ENOUGH ROOM.
C 
      IF (80 .LT. JINDX)  GO TO 1 
C 
C         IF THERE IS ENOUGH ROOM, PUT THE CHARACTER INTO THE BUFFER AND
C         INCREMENT THE CURRENT CHARACTER POSITION. 
C 
      JBUFF(JINDX)=JCHAR
      JINDX=JINDX+1 
C 
C         SET THE FLAG TO INDICATE ACCEPTANCE OF THE CURRENT CHARACTER
C         AND RETURN TO THE CALLER. 
C 
      JFLAG=0 
      RETURN
C 
C         THE LINE BUFFER HAS BEEN TERMINATED.  SET THE TERMINATING 
C         CHARACTER AND THE LIMIT INDEX.
C 
    1 JBUFF(JINDX)=-1 
      JLENG=JINDX 
C 
C         RESET THE USER INDEX FOR THE NEXT LINE. 
C 
      JINDX=1 
C 
C         SET THE FLAG TO INDICATE REJECTION OF THE CURRENT CHARACTER 
C         AND RETURN TO THE CALLER. 
C 
      JFLAG=1 
      RETURN
C 
      END 

File: waite/Cstage2/iwch.c
#include <stdio.h>

void
iwrch(jchar,jbuff,jindx,jflag,jleng)
int jchar,jbuff[BUFSIZ-1],*jindx,*jflag,*jleng;
/*
 * This routine enters a character into the line buffer.  If the
 * character is negative, it is taken to be a terminator.  If the line
 * buffer is full, it is also terminated and the character is ignored.
 * If the line buffer is terminated, a flag is returned to the user
 * indicating this fact.  The line buffer limit is set to index the
 * position beyond the last significant character.  If the line buffer
 * is not terminated, then the character is entered at the indexed
 * position and the index is incremented.  The limit index is not
 * altered in this case.
 *
 *        jchar  is the integer representing the character
 *
 *        jbuff  is the address of the line buffer
 *
 *        jindx  is the index to the line buffer
 *
 *        jflag is set to zero if the line buffer is not full and the
 *             character inserted was not cr/lf.  It is set to 1
 *             otherwise
 *
 *       jleng is the last+1 index of information in the buffer
 */
{
/*
 *        If the character is negative, the buffer must be terminated.
 *        It is immaterial whether the buffer is full or not.
 */
	if (jchar >= 0 &&
/*
 *       If the user is attempting to put a significant character into
 *        the line buffer, we must ensure that there is enough room.
 */
		BUFSIZ-2 >= *jindx) {
/*
 *        If there is enough room, put the character into the buffer and
 *        increment the current character position.
 */
		jbuff[*jindx]=jchar;
		*jindx = *jindx+1;
/*
 *        Set the flag to indicate acceptance of the current character
 *        and return to the caller.
 */
		*jflag=0;
	} else {
/*
 *        The line buffer has been terminated.  Set the terminating
 *        character and the limit index.
 */
		jbuff[*jindx] = -1;
		*jleng = *jindx;
/*
 *        Reset the user index for the next line.
 */
		*jindx=1;
/*
 *        Set the flag to indicate rejection of the current character
 *        and return to the caller.
 */
		*jflag=1;
	}
}
